### PR TITLE
MAINT: Use list and dict comprehension when possible

### PR DESCRIPTION
**Pull request by rth**
This uses list comprehension when possible (instead of creating a list by appending to it in a for loop), which is faster and less verbose.

Tried to only make the changes in some simple cases when it doesn't hurt readability.

### CODE REVIEW

File: numpy/core/_dtype.py
Comment (eric-wieser) [inline numpy/core/_dtype.py]: This is actually different, and produces tuples instead of lists. I'm not sure if it matters, but it's worth noting that it is different.

File: numpy/f2py/crackfortran.py
Comment (eric-wieser) [inline numpy/f2py/crackfortran.py]: This is not safe - this line is deliberately copying the keys, as it modifies them during iteration.

Comment (rth) [reply numpy/f2py/crackfortran.py]: Absolutely -- reverted. Thanks.

**Comment by rth** [comment]
Also added some `dict` comprehension where it helped readability. CI is green now.

File: numpy/core/tests/test_einsum.py
Comment (eric-wieser) [inline numpy/core/tests/test_einsum.py]: nit: Would be easier to read if you swapped the order of the arguments to `zip`

File: numpy/core/tests/test_numerictypes.py
Comment (eric-wieser) [inline numpy/core/tests/test_numerictypes.py]: While you're changing this line, may as well replace `item[1]` with `dtype` from line 77

File: numpy/f2py/common_rules.py
Comment (eric-wieser) [inline numpy/f2py/common_rules.py]: While you're here - changing the outer loop to iterate over `block['common'].items()` would make this loop easier to read

**Comment by rth** [comment]
Thanks @eric-wieser! Addressed your comments.

**Review by eric-wieser — APPROVED** [review]
Looks good to me now. Will leave this for someone else to double check.

Would be good to squash these commits, but we can do that from the github UI, so no need for you to do it from your end

File: numpy/core/tests/test_einsum.py
Comment (shoyer) [inline numpy/core/tests/test_einsum.py]: Could just be `dict(zip(chars, sizes))`

File: numpy/core/tests/test_numerictypes.py
Comment (shoyer) [inline numpy/core/tests/test_numerictypes.py]: Are you sure this is equivalent? Previously we iterated through each element of `dtype` -- now we don't.

File: numpy/f2py/common_rules.py
Comment (shoyer) [inline numpy/f2py/common_rules.py]: use `vars_` for variables with name conflicts with builtins

File: numpy/core/tests/test_numerictypes.py
Comment (eric-wieser) [reply numpy/core/tests/test_numerictypes.py]: Nope, we iterated through each item of `descr`. `dtype` is the name being used for the elements within the `descr`, set on line 77.

**Comment by rth** [comment]
Thanks for the review @shoyer ! I addressed your comments.

**Review by shoyer — APPROVED** [review]

File: numpy/core/_dtype.py
Comment (rth) [inline numpy/core/_dtype.py]: Hmm, I may have forgotten to handle the case when `names` is an empty list, then the zip unpacking will fail.

Not sure if that can happen, or are we sure that `dtype.names` will never be empty?

**Comment by eric-wieser** [comment]
Thanks @rth, you've just given me the reason I needed to never use `zip(*x)` in place of a loop.

`names` can be an empty list, so you do need to change that part of the patch.

**Comment by rth** [comment]
Yes, I keep forgetting that one.
`zip(*x)` for unpacking is much faster, but I guess here the `len(names)` is too small for it to matter.

Reverted that change altogether.

**Comment by shoyer** [comment]
This change is probably good to go in now, but for what it's worth, these sort of small semantic differences are why I'm skeptical that pervasive cleanups like this are worthwhile. It's better to do clean-ups when you're actually modifying adjacent code, which ensures you're familiar with how it works and understand the level of test coverage.

**Comment by rth** [comment]
> This change is probably good to go in now,

Thanks!

> but for what it's worth, these sort of small semantic differences are why I'm skeptical that pervasive cleanups like this are worthwhile. It's better to do clean-ups when you're actually modifying the adjacent code, which ensures you're familiar with how it works and understand the level of test coverage.

I agree that there is a risk in any refactoring, but lately, I have come to appreciate simple mechanical fixes across all the code base. At least theoretically, focusing on one thing should be less error-prone (though I have not been very careful at the beginning of this PR -- thanks for reviewers for caching those mistakes). Also because it's a simple change, once the detection method is setup (here multi-line regexp) it doesn't too much extra effort to apply it on different projects. I find the corresponding feedback very instructive.

For instance, that issue that I almost introduced in https://github.com/numpy/numpy/pull/12445#discussion_r236836095 was detected because I made a similar PR to matplotlib and a maintainer there detected it (after merge unfortunately). In any case, it's a more immediate way of sharing knowledge between projects. 

Also now that this issue with zip unpacking was raised (probably for the nth time), one could look how this pattern is used in different projects (see [zip_unpacking_occurences.txt](https://github.com/numpy/numpy/files/2626927/zip_unpacking_occurences.txt))  and wonder whether they all handle the case of empty iterable properly..

File: numpy/distutils/npy_pkg_config.py
Comment (charris) [inline numpy/distutils/npy_pkg_config.py]: This looks like a copy.

File: numpy/compat/_inspect.py
Comment (jaimefrio) [inline numpy/compat/_inspect.py]: Maybe I'm missing something, but do you really need the index here? Why not:
```
specs = [strseq(arg, ...) for arg in args]
```

Comment (rth) [reply numpy/compat/_inspect.py]: You are right -- simplified it more.

**Comment by charris** [comment]
Thanks @rth .

**Pull request merged**
